retry_decorator/retry_decorator.py

```diff
@@ -3,54 +3,79 @@
 # Copyright: Patrick Ng - 2012
 #
 
-from __future__ import print_function
-
-import traceback
 import logging
 import time
 import random
-import sys
 
 
-def retry(ExceptionToCheck, tries=10, timeout_secs=1.0, logger=None, callback_by_exception=None):
+def _deco_retry(f, exception_to_retry, tries, timeout_secs, logger, callback_by_exception):
+    def f_retry(*args, **kwargs):
+        mtries, mdelay = tries, timeout_secs
+        run_one_last_time = True
+        while mtries > 1:
+            try:
+                return f(*args, **kwargs)
+            except exception_to_retry as e:
+                # check if this exception is something the caller wants special handling for
+                callback_errors = callback_by_exception or {}
+                for error_type in callback_errors:
+                    if isinstance(e, error_type):
+                        callback_logic = callback_by_exception[error_type]
+                        should_break_out = run_one_last_time = False
+                        if isinstance(callback_logic, (list, tuple)):
+                            callback_logic, should_break_out = callback_logic
+                            if isinstance(should_break_out, (list, tuple)):
+                                should_break_out, run_one_last_time = should_break_out
+                        callback_logic()
+                        if should_break_out:  # caller requests we stop handling this exception
+                            break
+                half_interval = mdelay * 0.10  # interval size
+                actual_delay = random.uniform(mdelay - half_interval, mdelay + half_interval)
+                msg = "Retrying in %.2f seconds ..." % actual_delay
+                logging_object = logger or logging
+                logging_object.exception(msg)
+                time.sleep(actual_delay)
+                mtries -= 1
+                mdelay *= 2
+        if run_one_last_time:  # one exception may be all the caller wanted in certain cases
+            return f(*args, **kwargs)
+
+    return f_retry  # true decorator
+
+
+def retry(exception_to_retry=Exception, tries=10, timeout_secs=1.0, logger=None, callback_by_exception=None):
```

Comment:
The `exception_to_retry` sounds like it would retry an exception. What do you think about `retry_on_exception` or maybe simply `exc`?

Reply:
great idea. i usually prefer more verbose names, but definitely open to those suggestions. will think a bit on it and change when i have a moment.

Reply:
ok, changed to `exc`.